In [1]:
%matplotlib inline

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Problem 3. Data cleaning and preprocessing (1 point)
This should be self-explanatory. In an appropriate notebook (probably different than your previous one), explore different ways to clean and preprocess the dataset.

This is still part of your research. That is, don't be afraid to _try out different approaches to the same problem_. E.g., if you have a lot of missing values, you may not know right away how to handle them. Experimenting with several approaches will give you a better indication what works well for your data and goals.

In [4]:
asthma_disease_data = pd.read_csv('../data/asthma_disease_data.csv')

In [5]:
asthma_disease_data.head()

PatientID  Age  Gender  Ethnicity  EducationLevel        BMI  Smoking  \
0       5034   63       0          1               0  15.848744        0   
1       5035   26       1          2               2  22.757042        0   
2       5036   57       0          2               1  18.395396        0   
3       5037   40       1          2               1  38.515278        0   
4       5038   61       0          0               3  19.283802        0   

   PhysicalActivity  DietQuality  SleepQuality  ...  LungFunctionFEV1  \
0          0.894448     5.488696      8.701003  ...          1.369051   
1          5.897329     6.341014      5.153966  ...          2.197767   
2          6.739367     9.196237      6.840647  ...          1.698011   
3          1.404503     5.826532      4.253036  ...          3.032037   
4          4.604493     3.127048      9.625799  ...          3.470589   

   LungFunctionFVC  Wheezing  ShortnessOfBreath  ChestTightness  Coughing  \
0         4.941206         0                  0               1         0   
1         1.702393         1                  0               0         1   
2         5.022553         1                  1               1         0   
3         2.300159         1                  0               1         1   
4         3.067944         1                  1               1         0   

   NighttimeSymptoms  ExerciseInduced  Diagnosis  DoctorInCharge  
0                  0                1          0       Dr_Confid  
1                  1                1          0       Dr_Confid  
2                  1                1          0       Dr_Confid  
3                  1                0          0       Dr_Confid  
4                  0                1          0       Dr_Confid  

[5 rows x 29 columns]

2. Ensure Data Validity

Check for common issues such as missing values, duplicates, and data types

In [7]:
asthma_disease_data.shape

(2392, 29)

In [8]:
asthma_disease_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   PatientID               2392 non-null   int64  
 1   Age                     2392 non-null   int64  
 2   Gender                  2392 non-null   int64  
 3   Ethnicity               2392 non-null   int64  
 4   EducationLevel          2392 non-null   int64  
 5   BMI                     2392 non-null   float64
 6   Smoking                 2392 non-null   int64  
 7   PhysicalActivity        2392 non-null   float64
 8   DietQuality             2392 non-null   float64
 9   SleepQuality            2392 non-null   float64
 10  PollutionExposure       2392 non-null   float64
 11  PollenExposure          2392 non-null   float64
 12  DustExposure            2392 non-null   float64
 13  PetAllergy              2392 non-null   int64  
 14  FamilyHistoryAsthma     2392 non-null   

In [9]:
asthma_disease_data.isnull().sum()

PatientID                 0
Age                       0
Gender                    0
Ethnicity                 0
EducationLevel            0
BMI                       0
Smoking                   0
PhysicalActivity          0
DietQuality               0
SleepQuality              0
PollutionExposure         0
PollenExposure            0
DustExposure              0
PetAllergy                0
FamilyHistoryAsthma       0
HistoryOfAllergies        0
Eczema                    0
HayFever                  0
GastroesophagealReflux    0
LungFunctionFEV1          0
LungFunctionFVC           0
Wheezing                  0
ShortnessOfBreath         0
ChestTightness            0
Coughing                  0
NighttimeSymptoms         0
ExerciseInduced           0
Diagnosis                 0
DoctorInCharge            0
dtype: int64

In [10]:
asthma_disease_data.duplicated().sum()

0

The data is valid - no missing values, no duplicated rows.
The dataset has 2392 observations and 29 features.
Before starting some needed transofmations I will make a copy of the original dataset.

In [12]:
asthma_disease_data_cleaned = asthma_disease_data.copy()

Colimns rename

In [13]:
asthma_disease_data.columns

Index(['PatientID', 'Age', 'Gender', 'Ethnicity', 'EducationLevel', 'BMI',
       'Smoking', 'PhysicalActivity', 'DietQuality', 'SleepQuality',
       'PollutionExposure', 'PollenExposure', 'DustExposure', 'PetAllergy',
       'FamilyHistoryAsthma', 'HistoryOfAllergies', 'Eczema', 'HayFever',
       'GastroesophagealReflux', 'LungFunctionFEV1', 'LungFunctionFVC',
       'Wheezing', 'ShortnessOfBreath', 'ChestTightness', 'Coughing',
       'NighttimeSymptoms', 'ExerciseInduced', 'Diagnosis', 'DoctorInCharge'],
      dtype='object')

In [14]:
asthma_disease_data_cleaned.columns = (
    asthma_disease_data_cleaned.columns
    .str.replace(r'(?<!^)(?=[A-Z])', '_', regex=True)  # Convert camelCase to snake_case
    .str.lower()  # Convert to lowercase
    .str.replace('i_d', 'id')  # Replace 'i_d' with 'id'
    .str.replace('b_m_i', 'bmi')  # Replace 'b_m_i' with 'bmi'
    .str.replace('f_e_v1', 'fev1')  # Replace 'f_e_v1' with 'fev1'
    .str.replace('f_v_c', 'fvc')  # Replace 'f_v_c' with 'fvc'
)

In [15]:
asthma_disease_data_cleaned.columns

Index(['patient_id', 'age', 'gender', 'ethnicity', 'education_level', 'bmi',
       'smoking', 'physical_activity', 'diet_quality', 'sleep_quality',
       'pollution_exposure', 'pollen_exposure', 'dust_exposure', 'pet_allergy',
       'family_history_asthma', 'history_of_allergies', 'eczema', 'hay_fever',
       'gastroesophageal_reflux', 'lung_function_fev1', 'lung_function_fvc',
       'wheezing', 'shortness_of_breath', 'chest_tightness', 'coughing',
       'nighttime_symptoms', 'exercise_induced', 'diagnosis',
       'doctor_in_charge'],
      dtype='object')

In [16]:
asthma_disease_data_cleaned

patient_id  age  gender  ethnicity  education_level        bmi  smoking  \
0           5034   63       0          1                0  15.848744        0   
1           5035   26       1          2                2  22.757042        0   
2           5036   57       0          2                1  18.395396        0   
3           5037   40       1          2                1  38.515278        0   
4           5038   61       0          0                3  19.283802        0   
...          ...  ...     ...        ...              ...        ...      ...   
2387        7421   43       1          0                2  29.059613        0   
2388        7422   18       1          0                1  20.740850        0   
2389        7423   54       0          3                2  37.079560        0   
2390        7424   46       1          0                2  23.444712        0   
2391        7425   26       1          0                0  28.123021        1   

      physical_activity  diet_quality  sleep_quality  ...  lung_function_fev1  \
0              0.894448      5.488696       8.701003  ...            1.369051   
1              5.897329      6.341014       5.153966  ...            2.197767   
2              6.739367      9.196237       6.840647  ...            1.698011   
3              1.404503      5.826532       4.253036  ...            3.032037   
4              4.604493      3.127048       9.625799  ...            3.470589   
...                 ...           ...            ...  ...                 ...   
2387           3.019854      6.119637       8.300960  ...            3.125249   
2388           5.805180      4.386992       7.731192  ...            1.132977   
2389           4.735169      8.214064       7.483521  ...            1.685962   
2390           9.672637      7.362861       6.717272  ...            3.481549   
2391           1.613138      7.412878       8.512253  ...            2.280613   

      lung_function_fvc  wheezing  shortness_of_breath  chest_tightness  \
0              4.941206         0                    0                1   
1              1.702393         1                    0                0   
2              5.022553         1                    1                1   
3              2.300159         1                    0                1   
4              3.067944         1                    1                1   
...                 ...       ...                  ...              ...   
2387           5.166032         0                    1                0   
2388           5.509502         0                    0                0   
2389           3.346877         1                    0                1   
2390           1.713274         0                    1                1   
2391           2.453284         1                    1                0   

      coughing  nighttime_symptoms  exercise_induced  diagnosis  \
0            0                   0                 1          0   
1            1                   1                 1          0   
2            0                   1                 1          0   
3            1                   1                 0          0   
4            0                   0                 1          0   
...        ...                 ...               ...        ...   
2387         0                   0                 1          1   
2388         1                   1                 0          1   
2389         1                   0                 1          1   
2390         0                   1                 1          0   
2391         0                   1                 0          0   

      doctor_in_charge  
0            Dr_Confid  
1            Dr_Confid  
2            Dr_Confid  
3            Dr_Confid  
4            Dr_Confid  
...                ...  
2387         Dr_Confid  
2388         Dr_Confid  
2389         Dr_Confid  
2390         Dr_Confid  
2391         Dr_Confid  

[2392 rows x 29 columns]

The data is cleaned, every feature has the right type and the dataset is ready for analysis.